design-document.md

```diff
@@ -14,4 +24,169 @@
+follows:
+
+1. The user runs pf := Prove(x). It splits x and pf into {x} and {pf}
+   respectively and sends ({x:i}, {pf:i}) to aggregator i.
+1. Each aggregator i runs {vf:i} := Query({x:i}, {pf:i}, r) ands sends {vf:i} to
+   the leader.
+1. The leader runs b := Decide({vf}, r) and sends b to each of the aggregators.
+
+If b=1, then each aggregator i adds its input share {x:i} into its share of the
+aggregate. Once all of the inputs have been validated and aggregated, the
+aggregators send their aggregate shares to the leader, who adds them together to
+obtain the final result.
+
+**Proof generation and verification.**
+[[TODO: Describe how to construct proof systems for languages recognized by
+validity circuits with G-gates, a la [BCC+19, Theorem 4.3].]]
+
+### Changes to the core protocol
+
+**Coordinating state.**
+The state of the input-validation protocol is maintained by the leader; except
+for aggregation of the input shares, the other aggregators are completely
+stateless. This is accomplished by making the following changes to the core
+protocol.
+
+1. The user sends all of its shares to the leader. To maintain privacy, the user
+   encrypts each input/proof share under the public key of the share's
+   recipient.
+1. The leader forwards each encrypted share to its intended recipient. Each
+   aggregator decrypts its input and proof share, computes its verification
+   share, and sends its verification share to the aggregator as usual.
+1. If b=1 in the last step, then the leader also sends along the encrypted input
+   share to each aggregator so that they can decrypt and aggregate the share
+   without needing to maintain state from the previous step.
+
+**Minimizing bandwdith overhead.**
+In most linear secret sharing schemes, the length of each share is equal to the
+length of the input. Therefore, the communication overhead for the user is
+O(s\*(n+p(n))). This can be reduced to O(s+n+p(n)) with the following standard
+trick.
+
+Let x be an element of K^n for some n. Suppose we split in into {x} by choosing
+{x:1}, ..., {x:s-1} at random and letting {x:s} = in - ({x:1} + ... + {x:s-1}).
+We could instead choose s-1 random seeds k[s-1], ..., k[s-1] for a pseudorandom
+number generator PRG and let {x:i} = PRG(k[i], n) for each i. This effectively
+"compresses" s-1 of the shares to O(1) space.
+
+### Primitives
+
+This section describes the core cryptographic primitives of the system.
+
+#### Finite field arithmetic
+
+The algorithms that comprise the input-validation protocol --- Prove, Query, and
+Decide --- are constructed by generating and evaluating polynomials over a
+finite field. As such, the main ingredient of Prio is an implementation of
+arithmetic in a finite field suitable for the given application.
+
+We will use a prime field. The choice of prime is influenced by the following
+criteria:
+
+1. **Field size.** How big the field needs to be depends on the type of data
```

Comment:
Does arbitrary sized finite fields imply the use of bignums? Will the field sizes be carefully chosen for performance (e.g. recommended EC primes of 256, 384, etc)?

Reply:
I don't think we should try to support arbitrary fields; instead, we should pick a few fields of different sizes and develop optimized versions of each. I suspect that generic bignum libraries would be too slow. (That said, a coworker of mine developed some pretty fast code that works for any prime field of size < 2^126. We'll be open sourcing this very soon.) As far as using off-the-shelf code for ECC fields. My sense is that he design criteria for ECC are quite different from ours. Hence, I'm not sure that this would be suitable. Worth looking into, though.